crates/xtask-bump-check/src/xtask.rs

```diff
@@ -0,0 +1,338 @@
+/// * The maximum version is for beta channel, and the second one is for stable.
+fn beta_and_stable_branch(repo: &git2::Repository) -> CargoResult<[git2::Branch<'_>; 2]> {
+    let mut release_branches = Vec::new();
+    for branch in repo.branches(Some(git2::BranchType::Remote))? {
+        let (branch, _) = branch?;
+        let Some(version) = branch.name()?.and_then(|n| n.strip_prefix("origin/rust-")) else {
+            continue;
+        };
+        release_branches.push((version.to_semver()?, branch));
+    }
+    release_branches.sort_unstable_by(|a, b| a.0.cmp(&b.0));
+
+    let beta = release_branches.pop().unwrap();
+    let stable = release_branches.pop().unwrap();
+
+    assert_eq!(beta.0.major, 1);
+    assert_eq!(beta.0.patch, 0);
+    assert_eq!(stable.0.major, 1);
+    assert_eq!(stable.0.patch, 0);
+    assert_ne!(beta.0.minor, stable.0.minor);
+
+    Ok([beta.1, stable.1])
+}
+
+/// Gets the referenced commit to compare if version bump needed.
+///
+/// * When merging into nightly, check the version with beta branch
+/// * When merging into beta, check the version with stable branch
+/// * When merging into stable, check against crates.io registry directly
+fn get_referenced_commit<'a>(
+    repo: &'a git2::Repository,
+    base: &git2::Commit<'a>,
+) -> CargoResult<Option<git2::Commit<'a>>> {
+    let [beta, stable] = beta_and_stable_branch(&repo)?;
+    let rev_id = base.id();
+    let stable_commit = stable.get().peel_to_commit()?;
+    let beta_commit = beta.get().peel_to_commit()?;
+
+    let commit = if rev_id == stable_commit.id() {
+        None
+    } else if rev_id == beta_commit.id() {
+        Some(stable_commit)
+    } else {
+        Some(beta_commit)
+    };
+
+    Ok(commit)
+}
+
+/// Lists all changed workspace members between two commits.
+///
+/// Assumption: Paths of workspace members. See `member_dirs`.
+fn changed<'r, 'ws>(
+    ws: &'ws Workspace<'_>,
+    repo: &'r git2::Repository,
+    base_commit: &git2::Commit<'r>,
+    head: &git2::Commit<'r>,
+) -> CargoResult<HashMap<&'ws str, &'ws Package>> {
+    let member_dirs = ["crates", "credential", "benches"];
+    let ws_members: HashMap<&str, &Package> =
+        HashMap::from_iter(ws.members().map(|m| (m.name().as_str(), m)));
+    let base_tree = base_commit.as_object().peel_to_tree()?;
+    let head_tree = head.as_object().peel_to_tree()?;
+    let diff = repo
+        .diff_tree_to_tree(Some(&base_tree), Some(&head_tree), Default::default())
+        .unwrap();
+
+    let mut changed_members = HashMap::new();
+
+    for delta in diff.deltas() {
+        let mut insert_changed = |path: &Path| {
+            if !member_dirs.iter().any(|dir| path.starts_with(dir)) {
+                return;
+            }
+            let Some(member) = path.components().nth(1) else {
+                log::trace!("skipping {path:?}, not a valid member path");
+                return;
+            };
```

Comment:
Can we instead search for the entry where a `strip_prefix` works and then we take the next path element after that?

That would harden this code against someone adding a `member_dirs` that isn't just one level